File: app/src/main/java/com/ddu/ui/dialog/VerificationCodeDialog.java
package com.ddu.ui.dialog;

import android.content.Context;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ddu.R;
import com.ddu.icore.ui.view.NumberInputView;
import com.ddu.icore.util.AnimatorUtils;
import com.ddu.icore.util.sys.ViewUtils;

import static com.ddu.R.id.tv_error_msg;
import static com.ddu.R.id.tv_phone_number;

/**
 * Created by yzbzz on 2017/5/26.
 */

public class VerificationCodeDialog extends DialogFragment implements View.OnClickListener {

    public static final String EXTRA_PHONE = "extra_phone";

    private static final int COUNT_DOWN_TIME = 60 * 1000;
    private static final long COUNT_DOWN_INTERVAL = 1000;

    private TextView mTvPhoneNumber;
    private TextView mTvCountDown;
    private TextView mTvResend;
    private TextView mTvErrorMsg;
    private LinearLayout mLlVoiceCode;
    private NumberInputView mEtPhoneNumber;

    private RegisterTimer mRegisterTimer;

    private String mPhone = "";

    private Context mContext;

    public static VerificationCodeDialog newInstance(String phone) {
        Bundle args = new Bundle();
        args.putString(EXTRA_PHONE, phone);
        VerificationCodeDialog fragment = new VerificationCodeDialog();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPhone = getArguments().getString(EXTRA_PHONE);
        mRegisterTimer = new RegisterTimer(COUNT_DOWN_TIME, 250);
        mRegisterTimer.start();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mContext = context;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        getDialog().getWindow().setBackgroundDrawableResource(android.R.color.transparent);

        final LinearLayout mLlLogin = (LinearLayout) inflater.inflate(R.layout.fragment_verification_code, container, false);

        mTvErrorMsg = ViewUtils.findViewById(mLlLogin, tv_error_msg);
        mTvPhoneNumber = ViewUtils.findViewById(mLlLogin, tv_phone_number);
        mTvPhoneNumber.setText(mPhone);

        mLlVoiceCode = ViewUtils.findViewById(mLlLogin, R.id.ll_voice_code);
        mEtPhoneNumber = ViewUtils.findViewById(mLlLogin, R.id.et_phone_number);
        mEtPhoneNumber.setOnInputCallback(new NumberInputView.OnInputCallback() {
            @Override
            public void onInputComplete(String inputText) {
                mTvErrorMsg.setVisibility(View.VISIBLE);
                mLlLogin.startAnimation(AnimatorUtils.shake(mContext));
            }
        });
        mTvCountDown = ViewUtils.findViewById(mLlLogin, R.id.tv_count_down);
        mTvResend = ViewUtils.findViewById(mLlLogin, R.id.btn_resend);
        mTvResend.setOnClickListener(this);
        return mLlLogin;
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        switch (id) {
            case R.id.btn_resend:
                reSend();
                break;
        }
    }

    private void reSend() {
        mTvResend.setVisibility(View.GONE);
        mTvCountDown.setVisibility(View.VISIBLE);
        mRegisterTimer.start();
    }

    class RegisterTimer extends CountDownTimer {

        public RegisterTimer(long millisInFuture, long countDownInterval) {
            super(millisInFuture, countDownInterval);
        }

        @Override
        public void onFinish() {
            mTvResend.setVisibility(View.VISIBLE);
            mTvCountDown.setVisibility(View.GONE);
        }

        @Override
        public void onTick(long millisUntilFinished) {
            if (isAdded()) {
                mTvCountDown.setText(getResources().getString(R.string.format_count_down, millisUntilFinished / 1000));
                if (millisUntilFinished / 1000 == COUNT_DOWN_TIME / 1000 - 20) {
                    mLlVoiceCode.setVisibility(View.VISIBLE);
                }
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mRegisterTimer.cancel();
    }
}

File: app/src/main/java/com/ddu/ui/dialog/VerificationCodeDialog.kt
package com.ddu.ui.dialog

import android.content.Context
import android.os.Bundle
import android.os.CountDownTimer
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.LinearLayout
import android.widget.TextView
import com.ddu.R
import com.ddu.R.id.tv_error_msg
import com.ddu.R.id.tv_phone_number
import com.ddu.icore.common.ctx
import com.ddu.icore.common.loadAnimation
import com.ddu.icore.common.showKeyboard
import com.ddu.icore.ui.view.NumberInputView
import com.ddu.icore.util.sys.ViewUtils

/**
 * Created by yzbzz on 2017/5/26.
 */

class VerificationCodeDialog : DialogFragment(), View.OnClickListener {

    private var mTvPhoneNumber: TextView? = null
    private var mTvCountDown: TextView? = null
    private var mTvResend: TextView? = null
    private var mTvErrorMsg: TextView? = null
    private var mLlVoiceCode: LinearLayout? = null
    private var mEtPhoneNumber: NumberInputView? = null

    private var mRegisterTimer: RegisterTimer? = null

    private var mPhone: String? = ""

    private var mContext: Context? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPhone = arguments!!.getString(EXTRA_PHONE)
        mRegisterTimer = RegisterTimer(COUNT_DOWN_TIME.toLong(), 250)
        mRegisterTimer!!.start()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mContext = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialog.window!!.requestFeature(Window.FEATURE_NO_TITLE)
        dialog.window!!.setBackgroundDrawableResource(android.R.color.transparent)

        val mLlLogin = inflater.inflate(R.layout.fragment_verification_code, container, false) as LinearLayout

        mTvErrorMsg = ViewUtils.findViewById(mLlLogin, tv_error_msg)
        mTvPhoneNumber = ViewUtils.findViewById(mLlLogin, tv_phone_number)
        mTvPhoneNumber!!.text = mPhone

        mLlVoiceCode = ViewUtils.findViewById(mLlLogin, R.id.ll_voice_code)
        mEtPhoneNumber = ViewUtils.findViewById(mLlLogin, R.id.et_phone_number)
        mEtPhoneNumber!!.postDelayed({
            //                mEtPhoneNumber.getEt().performClick();
            //                mEtPhoneNumber.getll().performClick();
            activity?.showKeyboard(mEtPhoneNumber)
        }, 300)
        mEtPhoneNumber!!.setOnInputCallback {
            mTvErrorMsg!!.visibility = View.VISIBLE
            mLlLogin.startAnimation(ctx.loadAnimation(R.anim.shake))
        }
        mTvCountDown = ViewUtils.findViewById(mLlLogin, R.id.tv_count_down)
        mTvResend = ViewUtils.findViewById(mLlLogin, R.id.btn_resend)
        mTvResend!!.setOnClickListener(this)
        return mLlLogin
    }

    override fun onClick(v: View) {
        val id = v.id
        when (id) {
            R.id.btn_resend -> {
                mEtPhoneNumber!!.clearText()
                reSend()
            }
        }
    }

    private fun reSend() {
        mTvResend!!.visibility = View.GONE
        mTvCountDown!!.visibility = View.VISIBLE
        mRegisterTimer!!.start()
    }

    internal inner class RegisterTimer(millisInFuture: Long, countDownInterval: Long) : CountDownTimer(millisInFuture, countDownInterval) {

        override fun onFinish() {
            mTvResend!!.visibility = View.VISIBLE
            mTvCountDown!!.visibility = View.GONE
        }

        override fun onTick(millisUntilFinished: Long) {
            if (isAdded) {
                mTvCountDown!!.text = resources.getString(R.string.format_count_down, millisUntilFinished / 1000)
                if (millisUntilFinished / 1000 == (COUNT_DOWN_TIME / 1000 - 20).toLong()) {
                    mLlVoiceCode!!.visibility = View.VISIBLE
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mRegisterTimer!!.cancel()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            mEtPhoneNumber!!.performClick()
        }
    }

    companion object {

        val EXTRA_PHONE = "extra_phone"

        private val COUNT_DOWN_TIME = 60 * 1000
        private val COUNT_DOWN_INTERVAL: Long = 1000

        fun newInstance(phone: String): VerificationCodeDialog {
            val args = Bundle()
            args.putString(EXTRA_PHONE, phone)
            val fragment = VerificationCodeDialog()
            fragment.arguments = args
            return fragment
        }
    }
}
